src/attached_probe.h

```diff
@@ -40,84 +40,42 @@ class AttachedProbe {
                                                      std::optional<int> pid,
                                                      BPFtrace &bpftrace,
                                                      bool safe_mode = true);
-
-  ~AttachedProbe();
+  virtual ~AttachedProbe();
   AttachedProbe(const AttachedProbe &) = delete;
   AttachedProbe &operator=(const AttachedProbe &) = delete;
 
-  const Probe &probe() const;
-  int progfd() const;
-  int linkfd_ = -1;
+private:
+  AttachedProbe(int progfd, std::function<void()> cleanup);
+  const int progfd_;
+  const std::function<void()> cleanup_;
+  friend class AttachedIterProbe;
+  friend class AttachedWatchpointProbe;
+};
+
+class AttachedIterProbe : public AttachedProbe {
+public:
+  ~AttachedIterProbe() override = default;
+
+  const int iter_link_fd;
```

Comment:
For the kprobe nmissed work, we'll need to access the link fd. This is generally applicable to all probe types, so exposing a getter for link fd is reasonable and useful. But to do that with your proposed design/refactor, we'll need to create classes for all probe types, no? So see overall comment.